Allow vtkMRMLModelNode inputs in MarkupsToModel

tavaughan opened the pull request:
Renamed from #5

tavaughan committed:
Separated implementation of helper namespaces from declaration.

tavaughan committed:
STY: Much refactoring including vtk classes as per #2.

tavaughan committed:
Correct error messages on opening MarkupsToModel, see #6

tavaughan committed:
ENH Added functionality to take model nodes as input #3

tavaughan committed:
STY: Fixed numerous small issues in MarkupsToModel

- Also avoid a crash when input and output nodes are the same
- Related to #3 and SlicerIGT/SlicerIGT#138

tavaughan committed:
BUG: Re-added deprecated methods

lassoan commented on MarkupsToModel/Logic/vtkSlicerMarkupsToModelLogic.h:
```diff
@@ -53,30 +53,32 @@ class VTK_SLICER_MARKUPSTOMODEL_MODULE_LOGIC_EXPORT vtkSlicerMarkupsToModelLogic
+  void SetMarkupsNode( vtkMRMLMarkupsFiducialNode* newMarkups, vtkMRMLMarkupsToModelNode* moduleNode );
+
+  // DEPRECATED
+  bool UpdateClosedSurfaceModel( vtkMRMLMarkupsFiducialNode* markupsNode, vtkMRMLModelNode* modelNode,
```
Please don't deprecate UpdateClosedSurfaceModel and UpdateOutputCurveModel methods. We need these. See for example this application:
https://github.com/lassoan/SlicerSegmentEditorExtraEffects/blob/7b932ec64f49d82448e802bf7c6ef16079ad5b2f/SegmentEditorSurfaceCut/SegmentEditorSurfaceCutLib/SegmentEditorEffect.py#L377

tavaughan replied:
The replacement code in that case would be:

```
controlPoints = vtk.vtkPoints()
slicer.vtkCreateModelUtil.vtkMRMLNodeToVtkPoints( inputMarkup, controlPoints )
polyData = vtk.vtkPolyData()
slicer.vtkCreateCurveUtil.GenerateLinearCurveModel( controlPoints, polyData )
outputModel.SetAndObservePolyData( polyData )
```

So it's a bit longer than I originally forecasted. Wonder if it's worth simplifying. What do you think?

tavaughan replied:
(Actually, having thought a bit about that code, I just realized how strange "LinearCurve" is. I think I'll change it to GenerateLineModel...)

EDIT: Nevermind, GenerateLineModel doesn't make much sense if connecting more than 2 points. I guess linear curve is still a better name, but I'm open to another suggestion.

lassoan replied:
It is expected that a module exposes its public API through its module logic and MRML classes. It is highly unusual to encourage developers to use a utility class directly.

vtkMRMLNodeToVtkPoints (name is not good, it should be just MRMLNodeToVtkPoints) functionality should be in the model and markup module (either in the MRML node or logic class).

It is not nice to pollute the namespace with a so class that has so generic name. At least the class name should start with the module name and instead of "util" it should be somewhat more specific (generator, converter, ...).

Having separate method for each interpolation type may be inconvenient for developers (e.g., to switch between different interpolator types, developers would need to add switch statements in their code).

In itself none of the above issues are significant, but overall it adds up to be an unusual and not very developer-friendly interface. Overall, while not ideal, I feel that simply not deprecating the existing interface in the logic would be a good solution. It would be the usual way to access module features and advanced developers could decide to use lower-level API if they prefer it. Maybe we simply overthink this, as in the long term things may be redesigned anyway.

tavaughan replied:
vtkMRMLNodeToVtkPoints => I don't understand your comment. What do you mean by "functionality should be in the model and markup module"? Which module(s) are you referring to?

Class names => I followed the convention of your earlier header files. It seems like we can have either one namespace with all our methods (which you changed, I assume because it was too many methods), or we can have a few different namespaces with fewer methods each (which has the issues you just described). Either way is non-ideal for a different reason. Please let me know how to proceed.

separate method for each interpolation type => I don't understand. In what use case would a developer be switching between these?

"simply not deprecating the existing interface in the logic would be a good solution" => Consider it done, then.

tavaughan committed:
BUG: Better error checking

tavaughan committed:
BUG: Fix VS2013 compile issue C2864

tavaughan committed:
BUG: Added logic classes to export

lassoan commented:
> vtkMRMLNodeToVtkPoints => I don't understand your comment. What do you mean by "functionality should be in the model and markup module"? Which module(s) are you referring to?

These kinds of utility functions that allows you to set content of a MRML node from a VTK object should be available in the module that defines these nodes (Models and Markups modules).

> Class names => I followed the convention of your earlier header files.

Those header files were not visible in Python because they were not wrapped. They were just private, internal helper classes or namespaces. If we create something that shows up in the global Slicer namespace then we have to make sure that name is specific enough (for example, by starting/including the module name is a simple way of indicating where the class belongs).

> separate method for each interpolation type 

If I develop a module where I want to easily switch between linear and spline interpolation then it is simpler for me to just change a variable value to switch between the interpolations than having some (simple) logic that calls different methods.

tavaughan commented:
> These kinds of utility functions that allows you to set content of a MRML node from a VTK object should be available in the module that defines these nodes (Models and Markups modules).

So these are changes for Slicer core, then?

> If we create something that shows up in the global Slicer namespace then we have to make sure that name is specific enough (for example, by starting/including the module name is a simple way of indicating where the class belongs).

Ok, so if there were names like vtkMRMLMarkupsToModelCurveGenerator, or vtkMRMLMarkupsToModelClosedSurfaceGenerator, those would be fine?

> If I develop a module where I want to easily switch between linear and spline interpolation then it is simpler for me to just change a variable value to switch between the interpolations than having some (simple) logic that calls different methods.

We decided above to expose the int variable in the logic member, so they'll be able to do that.

lassoan commented:
Yes to all three questions above.

lassoan commented:
Thanks for all your work. I hope I don't discourage you with my comments/suggestions. You've implemented great features!

tavaughan commented:
Not at all! I'm learning a lot and enjoying it. :) Besides, your points of improvement seem for the most part like very minor work. ;)

tavaughan commented:
> set content of a MRML node from a VTK object

I'm going the other way, VTK object from MRML node. Does this make a difference in terms of where the functionality should be implemented?

tavaughan committed:
STY: Changed helper class names to be more meaningful. Removed unwanted deprecated messages.

lassoan commented on MarkupsToModel/Logic/vtkSlicerMarkupsToModelCurveGeneration.h:
```diff
@@ -0,0 +1,93 @@
+      bool tubeLoop=vtkSlicerMarkupsToModelCurveGeneration::TUBE_LOOP_DEFAULT );
+
+    // Generates Kochanek Spline curve model.
+    static void GenerateKochanekCurveModel( vtkPoints* controlPoints, vtkPolyData* outputTubePolyData,
```
spline

lassoan commented on MarkupsToModel/Logic/vtkSlicerMarkupsToModelCurveGeneration.h:
```diff
@@ -0,0 +1,93 @@
+      vtkDoubleArray* markupsPointsParameters=NULL );
+
+    // Assign parameter values to points based on their position in the markups list (good for ordered point sets)
+    static void ComputePointParametersRawIndices(vtkPoints* controlPoints, vtkDoubleArray* outputPointParameters);
```
ComputeControlPointParametersFromIndices

tavaughan committed:
STY: Renamed helper functions to be more clear

tavaughan committed:
BUG: Point parameters from minimum spanning tree weren't being used

tavaughan committed:
BUG: Remove warning on making closed surface with only one point
(otherwise the user will get a meaningless warning when incrementally constructing a model point by point)

tavaughan commented:
Andras, you need to do the merge. I don't have write access to this repository.

lassoan commented:
Thank you!
